### PR TITLE
[GenAI] Add Langchain model server

### PR DESCRIPTION
**Pull request by ZeevRispler**
Add an option to serve model from Langchain using MLRun.

### CODE REVIEW

File: mlrun/frameworks/langchain/model_server.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    - abatch
+    in the future we will add streaming support.
+    Notice: In order to use this serving class, please ensure that the langchain_community package is installed.
+    """
+
+    def __init__(
+        self,
+        context: mlrun.MLClientCtx = None,
+        model_class: str = None,
+        llm: Union[str, LLM] = None,
```
Comment (guy1992l) [inline mlrun/frameworks/langchain/model_server.py:52]: I see `llm` can be a string as well, what would be the use case? Doesn't it replace the `model_class` in that case? If so, delete the `model_class`.

File: mlrun/frameworks/langchain/model_server.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    ):
+        """
+        Initialize a serving class for general llm usage.
+        :param context:           The mlrun context to use.
+        :param model_class:       The class of the model to use.
+        :param llm:               The llm object itself in case of local usage.
+        :param init_kwargs:       The initialization arguments to use while initializing the llm.
+        :param generation_kwargs: The generation arguments to use while generating text.
+        :param name:              The name of this server to be initialized.
+        :param model_path:        Not in use. When adding a model pass any string value
```
Comment (guy1992l) [inline mlrun/frameworks/langchain/model_server.py:66]: I guess we can "log" a langchain chain as a pickle and then the model path is the pickle file or model artifact. Write it as a TODO

File: mlrun/frameworks/langchain/model_server.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    """
+    LangChain Model serving class, inheriting the V2ModelServer class for being initialized automatically by the model
+    server and be able to run locally as part of a nuclio serverless function, or as part of a real-time pipeline.
+    For compatibility with langchain models, this class supports the following operations:
+    - invoke
+    - batch
+    - ainvoke
+    - abatch
+    in the future we will add streaming support.
+    Notice: In order to use this serving class, please ensure that the langchain_community package is installed.
```
Comment (guy1992l) [inline mlrun/frameworks/langchain/model_server.py]:
```suggestion
    LangChain Model serving class, inheriting the V2ModelServer class to be able to run locally or as part of a Nuclio serverless function.
    
    For compatibility with langchain models, this class supports the following methods:
    
    * invoke
    * batch
    * ainvoke
    * abatch
    * stream (TBD)
    * astream (TBD)

    This class can serve a LangChain chain or llm:
    
    * `LLM` - Expose a `langchain.LLM` object as a serverless function. Can work in offline and online modes:
      * Offline: Get an initialized `LLM` object - only in local testing (mock server)
      * Online: Get a class name and initialization keyword arguments to initialize the `LLM` object and serve it.
    * `Chain` (TBD) - Expose an entire `langchain` object as a pickled model logged to MLRun.

    > Notice: In order to use this serving class, please ensure that the `langchain` and `langchain_community` packages are installed.
```

File: mlrun/frameworks/langchain/model_server.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        self.llm = llm
+        self.init_kwargs = init_kwargs or {}
+        self.generation_kwargs = generation_kwargs
+
+    def load(self):
+        """
+        loads the model.
+        """
+        # If the llm is already an LLM object, use it directly
+        if isinstance(self.llm, LLM):
```
Comment (guy1992l) [inline mlrun/frameworks/langchain/model_server.py]: write a TODO to verify about the Chat objects of LangChain.

File: mlrun/frameworks/langchain/model_server.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            self.model = self.llm
+            return
+        # If the llm is a string (or not given, then we take default model), load the llm from langchain.
```
Comment (guy1992l) [inline mlrun/frameworks/langchain/model_server.py:83]: how you take a default model?

File: tests/frameworks/langchain/test_model_server.py [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+# 2. pull the desired model from the ollama repository (for example, in terminal: ollama pull llama3)
+# 3. run ollama with said model in terminal (for example, in terminal: ollama run llama3)
+def ollama_check_skip():
+    """
+    Check if ollama is installed
+    """
+    try:
+        result = subprocess.run(["ollama", "--help"], stdout=PIPE)
+    except Exception as e:
+        print(f"Error checking ollama: {e}")
```
Comment (guy1992l) [inline tests/frameworks/langchain/test_model_server.py:49]: why would it be raised?

**Review by guy1992l — CHANGES_REQUESTED** [review]
great stuff :)

Some questions and edit for the class docstring.

File: mlrun/frameworks/langchain/model_server.py (hunk 5 above)
Comment (ZeevRispler) [reply mlrun/frameworks/langchain/model_server.py:83]: you just init without specifing (ex: openai() and then you get gpt-3.5)

File: tests/frameworks/langchain/test_model_server.py (hunk 6 above)
Comment (ZeevRispler) [reply tests/frameworks/langchain/test_model_server.py:49]: didn't find command ollama (or something similar)

Comment (guy1992l) [reply tests/frameworks/langchain/test_model_server.py:49]: ah so True is to skip, ok!

File: mlrun/frameworks/langchain/model_server.py (hunk 2 above)
Comment (ZeevRispler) [reply mlrun/frameworks/langchain/model_server.py:66]: you want to do this in the model server?

File: mlrun/frameworks/langchain/model_server.py (hunk 4 above)
Comment (ZeevRispler) [reply mlrun/frameworks/langchain/model_server.py:71]: in addition to the LLM (which is an langchain class)? like a full chain or something?

**Pull request merged**